exporters/metrics/ocagent/src/main/java/io/opencensus/exporter/metrics/ocagent/OcAgentMetricsExporterWorker.java

```diff
@@ -0,0 +1,160 @@
+import io.grpc.ManagedChannelBuilder;
+import io.opencensus.common.Duration;
+import io.opencensus.metrics.export.Metric;
+import io.opencensus.metrics.export.MetricProducer;
+import io.opencensus.metrics.export.MetricProducerManager;
+import io.opencensus.proto.agent.metrics.v1.ExportMetricsServiceRequest;
+import io.opencensus.proto.agent.metrics.v1.MetricsServiceGrpc;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * Worker {@code Runnable} that polls Metric from Metrics library and batch export to Agent.
+ *
+ * <p>{@code OcAgentMetricsExporterWorker} will be started in a daemon {@code Thread}.
+ *
+ * <p>The state of this class should only be accessed from the thread which {@link
+ * OcAgentMetricsExporterWorker} resides in.
+ */
+@NotThreadSafe
+final class OcAgentMetricsExporterWorker implements Runnable {
+
+  private static final Logger logger =
+      Logger.getLogger(OcAgentMetricsExporterWorker.class.getName());
+
+  private final String endPoint;
+  private final boolean useInsecure;
+  private final long exportIntervalMillis;
+  private final long retryIntervalMillis;
+  private final String serviceName;
+  private final MetricProducerManager metricProducerManager;
+  private OcAgentMetricsServiceExportRpcHandler exportRpcHandler;
+  // private final Set<MetricDescriptor> registeredDescriptors = new HashSet<>();
+
+  OcAgentMetricsExporterWorker(
+      String endPoint,
+      boolean useInsecure,
+      Duration exportInterval,
+      Duration retryInterval,
+      String serviceName,
+      MetricProducerManager metricProducerManager) {
+    this.endPoint = endPoint;
+    this.useInsecure = useInsecure;
+    this.exportIntervalMillis = exportInterval.toMillis();
+    this.retryIntervalMillis = retryInterval.toMillis();
+    this.serviceName = serviceName;
+    this.metricProducerManager = metricProducerManager;
+  }
+
+  @Override
+  public void run() {
+    while (true) {
+      connect();
+      while (exportRpcHandler != null && !exportRpcHandler.isCompleted()) {
+        export();
+        sleep(exportIntervalMillis);
+      }
+      if (exportRpcHandler != null && exportRpcHandler.getTerminateStatus() != null) {
+        TerminateStatusRunnable runnable =
+            new TerminateStatusRunnable(exportRpcHandler.getTerminateStatus(), "Export");
+        new Thread(runnable).start();
+      }
+      sleep(retryIntervalMillis);
+    }
+  }
+
+  private void connect() {
+    ManagedChannelBuilder<?> channelBuilder = ManagedChannelBuilder.forTarget(endPoint);
+    if (useInsecure) {
+      channelBuilder.usePlaintext();
+    }
+    ManagedChannel channel = channelBuilder.build();
+    MetricsServiceGrpc.MetricsServiceStub stub = MetricsServiceGrpc.newStub(channel);
+    exportRpcHandler = OcAgentMetricsServiceExportRpcHandler.create(stub);
+    ExportMetricsServiceRequest firstRequest =
```

Comment:
what is the purpose of sending firstRequest without any metrics?

Reply:
The first message is used to open the RPC and provide initial metadata to Agent. E.g `Node` is needed only in the first message. After that RPC will be kept open and `Node` can be omitted in the subsequent messages.

Reply:
SG.

Reply:
I don't have privilege to approve it.